lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -2143,51 +2143,261 @@ namespace Js
+    class NumberFormatPartsBuilder
+    {
+    private:
+        double num;
+        const char16 *formatted;
+        const charcount_t formattedLength;
+        ScriptContext *sc;
+        UNumberFormatFields *fields;
+
+        static const UNumberFormatFields UnsetField = static_cast<UNumberFormatFields>(0xFFFFFFFF);
+
+        JavascriptString *GetPartTypeString(UNumberFormatFields field)
+        {
+            JavascriptLibrary *library = sc->GetLibrary();
+
+            if (field == UnsetField)
+            {
+                return library->GetIntlLiteralPartString();
+            }
+
+            switch (field)
+            {
+            case UNUM_INTEGER_FIELD:
+            {
+                if (JavascriptNumber::IsNan(num))
+                {
+                    return JavascriptString::NewCopySz(_u("nan"), sc);
+                }
+                else if (!NumberUtilities::IsFinite(num))
+                {
+                    return JavascriptString::NewCopySz(_u("infinity"), sc);
+                }
+                else
+                {
+                    return JavascriptString::NewCopySz(_u("integer"), sc);
+                }
+            }
+            case UNUM_FRACTION_FIELD: return JavascriptString::NewCopySz(_u("fraction"), sc);
+            case UNUM_DECIMAL_SEPARATOR_FIELD: return JavascriptString::NewCopySz(_u("decimal"), sc);
+
+            // The following three should only show up if UNUM_SCIENTIFIC is used, which Intl.NumberFormat doesn't currently use
+            case UNUM_EXPONENT_SYMBOL_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent symbol field");
+            case UNUM_EXPONENT_SIGN_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent sign field");
+            case UNUM_EXPONENT_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent field");
+
+            case UNUM_GROUPING_SEPARATOR_FIELD: return JavascriptString::NewCopySz(_u("group"), sc);
+            case UNUM_CURRENCY_FIELD: return JavascriptString::NewCopySz(_u("currency"), sc);
+            case UNUM_PERCENT_FIELD: return JavascriptString::NewCopySz(_u("percent"), sc);
+
+            // TODO(jahorto): Determine if this would ever be returned and what it would map to
+            case UNUM_PERMILL_FIELD: AssertOrFailFastMsg(false, "Unexpected permill field");
+
+            case UNUM_SIGN_FIELD: num < 0 ? JavascriptString::NewCopySz(_u("minusSign"), sc) : JavascriptString::NewCopySz(_u("plusSign"), sc);
+            default: AssertOrFailFastMsg(false, "Unexpected unknown part"); return nullptr;
+            }
+        }
+
+    public:
+        NumberFormatPartsBuilder(double num, const char16 *formatted, charcount_t formattedLength, ScriptContext *scriptContext)
+            : num(num)
+            , formatted(formatted)
+            , formattedLength(formattedLength)
+            , sc(scriptContext)
+            , fields(RecyclerNewArrayLeaf(sc->GetRecycler(), UNumberFormatFields, formattedLength))
```

Comment:
I'm surprised the Linux build didn't complain at you for not having `fields` marked as a `Field`, since it can hold a recycler pointer. Any idea why that would be?

Reply:
Oh, looks like the checker only applies that rule to classes that have ever been recycler-allocated themselves, and NumberFormatPartsBuilder is only created on the stack. Carry on.

---
In reply to: [186459626](https://github.com/Microsoft/ChakraCore/pull/5105#discussion_r186459626) [](ancestors = 186459626)

Reply:
I should probably still mark it as a Field in case this ever needs to be created in the heap in the future.

Reply:
I though the pointer itself would be a Field, not the thing being pointed at, like `Field(UNumberFormatFields *)`. Do I have this backwards?

---
In reply to: [186466305](https://github.com/Microsoft/ChakraCore/pull/5105#discussion_r186466305) [](ancestors = 186466305)

Reply:
I believe you're correct, all the `Field` specifiers on this class should contain the whole type

Reply:
Nope, you're right.